### PR TITLE
Decomposition of integration testsuite - proof of concept

### LINKED ISSUE #28012: Decomposition of dev/tests testsuites across corresponding modules
### Summary (*)

Magento 2.3 or 2.4. There are a lot of tests inside `dev/tests` directory, which violates modularity of platform, because while uninstalling (I mean completely removing) package from vendor tests stays in `dev/tests`. Wouldn't be better to spread them across corresponding modules? What if, I would like to uninstall module `Magento_Wishlist` and still run all integration tests? I still will have its tests in `dev/tests/integration/testsuite/Magento/Wishlist`, which won't work if I remove composer package with physical code.

### Examples (*)

Test inside `dev/tests/integration/testsuite/Magento/*`.

### Proposed solution

Copy-paste modules tests - I could start with integration ones into corresponding modules, so testsuite is decoupled from fixed modules set.

### PR DESCRIPTION
**Pull request by bartoszkubicki**
### Description (*)

All idea comes from this [issue](https://github.com/magento/magento2/issues/28012). @ihor-sviziev I took apart of `Magento_Rss`, `Magento_AdminNotification` also, as there is test here which contains also file data provider. Generally, it was very easy to move the test and locally all tests worked the same way. I don't see any problem in moving, and refactoring tests in similar way (optimizing imports, some return types, types hints)

### Fixed Issues (if relevant)

1. Fixes https://github.com/magento/magento2/issues/28012

### Manual testing scenarios (*)

1. Just run moved integration tests. If everything still works and tests pass it means that they were successfully decomposed from testsuite

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by orlangur** [comment]
> Just run magento integration tests. If everything still works and tests pass it means that they were successfully decomposed from testsuite

Small note: make sure decomposed tests where actually executed. You don't need to run all tests actually, just moved ones, CI will check the rest.

**Review by orlangur — CHANGES_REQUESTED** [review]
LGTM except I don't think we need to add visual noise into tests with `: void`.

Also, adding `declare strict types` should be a separate effort. 

In pull requests dedicated for integration tests modularity let's just move the tests. Can you do it all over the code at once with some reusable automation?

**Comment by ihor-sviziev** [comment]
> LGTM except I don't think we need to add visual noise into tests with `: void`.
> 
> Also, adding `declare strict types` should be a separate effort.
> 
> In pull requests dedicated for integration tests modularity let's just move the tests. Can you do it all over the code at once with some reusable automation?

That’s why I wanted to wait until php7.4 & phpunit 9 changes will be delivered to 2.4-develop branch, probably they should contains all these strict types and return void.

**Comment by bartoszkubicki** [comment]
@ihor-sviziev is there any problem between `php7.4 & phpunit 9` and strct types?
I don't know if it is easier to to these tasks iteratively - moving tests and fixing paths to fixtures AND later adding all types-like stuff. 

For sure there is no problem in dividing in particular PRs changes into commits like 1. moving tests 2. fixtures fixing 3. slight refactor + types

If all changes can be backward incompatible (for 2.4) I would say maybe it is better to take a chance of bigger changes. My recomendations is maybe not to bundle all work for all module into one PR, because it may be more difficult to review. Some smaller modules may be delivered together, but Magneto_Catalog should be done separately.

**Comment by ihor-sviziev** [comment]
@slavvka could you bring results of integration test failure? We can't see any failures in build report

**Comment by VladimirZaets** [comment]
@ihor-sviziev @bartoszkubicki , I reruned tests cause looks like something went wrong, we also haven't test results in our internal infrastructure

**Comment by ihor-sviziev** [comment]
@VladimirZaets the same result, integration tests fails.
Maybe some change on infra side should be changed in order to support integration tests in module directories

**Comment by VladimirZaets** [comment]
@ihor-sviziev @bartoszkubicki `testReferencesFromStaticFiles` test with different data set is failed. 
```
Due to LESS library specifics, the '../images/arrows-bg.svg' cannot be tested.
#0 /var/www/html/dev/tests/integration/testsuite/Magento/Test/Integrity/StaticFilesTest.php(115): PHPUnit\Framework\Assert::markTestSkipped('Due to LESS lib...')
#1 [internal function]: Magento\Test\Integrity\StaticFilesTest->testReferencesFromStaticFiles('adminhtml', 'Magento/backend', '', '', 'css/styles-old....', '/var/www/html/a...')
#2 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestCase.php(1071): ReflectionMethod->invokeArgs(Object(Magento\Test\Integrity\StaticFilesTest), Array)
#3 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestCase.php(939): PHPUnit\Framework\TestCase->runTest()
#4 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestResult.php(698): PHPUnit\Framework\TestCase->runBare()
#5 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestCase.php(894): PHPUnit\Framework\TestResult->run(Object(Magento\Test\Integrity\StaticFilesTest))
#6 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php(755): PHPUnit\Framework\TestCase->run(Object(PHPUnit\Framework\TestResult))
#7 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php(755): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#8 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php(755): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#9 /var/www/html/vendor/phpunit/phpunit/src/TextUI/TestRunner.php(545): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#10 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(195): PHPUnit\TextUI\TestRunner->doRun(Object(PHPUnit\Framework\TestSuite), Array, true)
#11 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(148): PHPUnit\TextUI\Command->run(Array, true)
#12 /var/www/html/vendor/phpunit/phpunit/phpunit(53): PHPUnit\TextUI\Command::main()
#13 {main}
```

I think we should try to run it locally

**Comment by bartoszkubicki** [comment]
I don't understand - what we should try to run locally

**Comment by ihor-sviziev** [comment]
@bartoszkubicki as I understood - following integration test fails: `Magento\Test\Integrity\StaticFilesTest->testReferencesFromStaticFiles`. Please try to run it locally, probably you'll see the same result

**Comment by bartoszkubicki** [comment]
@ihor-sviziev @VladimirZaets where is this test placed? I nor have it on my branch neither see it on 2.4-develop in github -. https://github.com/magento/magento2/tree/2.4-develop/dev/tests/static/testsuite/Magento/Test/Integrity

**Comment by ihor-sviziev** [comment]
> @ihor-sviziev @VladimirZaets where is this test placed? I nor have it on my branch neither see it on 2.4-develop in github -. https://github.com/magento/magento2/tree/2.4-develop/dev/tests/static/testsuite/Magento/Test/Integrity

Here is https://github.com/magento/magento2/blob/2.4-develop/dev/tests/integration/testsuite/Magento/Test/Integrity/StaticFilesTest.php

**Comment by bartoszkubicki** [comment]
@ihor-sviziev @VladimirZaets I have this test passed locally

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @bartoszkubicki,
Now changes from php 7.4 and phpunit 9 were merged. Could you fix conflict with 2.4-develop branch, so we'll see correct test results? Maybe fix for this issue was already delievered in 2.4-develop

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/AdminNotification/Test/Integration/Controller/Adminhtml/Notification/MarkAsReadTest.php [hunk 1]
```diff
@@ -1,21 +1,26 @@
...
  * @magentoAppArea adminhtml
  */
-class MarkAsReadTest extends \Magento\TestFramework\TestCase\AbstractBackendController
+class MarkAsReadTest extends AbstractBackendController
 {
     /**
      * @inheritdoc
      */
-    protected function setUp(): void
+    public function setUp(): void
```
Comment (ihor-sviziev) [inline app/code/Magento/AdminNotification/Test/Integration/Controller/Adminhtml/Notification/MarkAsReadTest.php]: Please revert it, it should be protected

File: app/code/Magento/AdminNotification/Test/Integration/Controller/Adminhtml/Notification/MassMarkAsReadTest.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+use Magento\Framework\Exception\AuthenticationException;
+use Magento\TestFramework\TestCase\AbstractBackendController;
+
+class MassMarkAsReadTest extends AbstractBackendController
+{
+    /**
+     * @return void
+     * @throws AuthenticationException
+     */
+    public function setUp(): void
```
Comment (ihor-sviziev) [inline app/code/Magento/AdminNotification/Test/Integration/Controller/Adminhtml/Notification/MassMarkAsReadTest.php]: Please revert it, it should be protected

```suggestion
    protected function setUp(): void
```

File: app/code/Magento/AdminNotification/Test/Integration/Controller/Adminhtml/Notification/MassRemoveTest.php [hunk 3]
```diff
@@ -1,15 +1,24 @@
...
 
-class MassRemoveTest extends \Magento\TestFramework\TestCase\AbstractBackendController
+class MassRemoveTest extends AbstractBackendController
 {
-    protected function setUp(): void
+    /**
+     * @return void
+     * @throws AuthenticationException
+     */
+    public function setUp(): void
```
Comment (ihor-sviziev) [inline app/code/Magento/AdminNotification/Test/Integration/Controller/Adminhtml/Notification/MassRemoveTest.php]: Please revert it, it should be protected

File: app/code/Magento/AdminNotification/Test/Integration/Controller/Adminhtml/Notification/RemoveTest.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+use Magento\Framework\Exception\AuthenticationException;
+use Magento\TestFramework\TestCase\AbstractBackendController;
+
+class RemoveTest extends AbstractBackendController
+{
+    /**
+     * @return void
+     * @throws AuthenticationException
+     */
+    public function setUp(): void
```
Comment (ihor-sviziev) [inline app/code/Magento/AdminNotification/Test/Integration/Controller/Adminhtml/Notification/RemoveTest.php]: Please revert it, it should be protected

File: app/code/Magento/AdminNotification/Test/Integration/Model/ResourceModel/Inbox/Collection/CriticalTest.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+use Magento\AdminNotification\Model\ResourceModel\Inbox\Collection\Critical;
+use Magento\TestFramework\Helper\Bootstrap;
+use PHPUnit\Framework\TestCase;
+
+class CriticalTest extends TestCase
+{
+    /**
+     * @var Critical
+     */
+    protected $model;
```
Comment (ihor-sviziev) [inline app/code/Magento/AdminNotification/Test/Integration/Model/ResourceModel/Inbox/Collection/CriticalTest.php]: Please make it private 
```suggestion
    private $model;
```

File: app/code/Magento/AdminNotification/Test/Integration/_files/notifications.php [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+use Magento\Framework\Notification\MessageInterface;
+use Magento\TestFramework\Helper\Bootstrap;
+
+$om = Bootstrap::getObjectManager();
+$message = $om->create(Inbox::class);
+$message->setSeverity(
+    MessageInterface::SEVERITY_CRITICAL
+)->setTitle(
+    'Unread Critical 1'
+)->save();
```
Comment (ihor-sviziev) [inline app/code/Magento/AdminNotification/Test/Integration/_files/notifications.php]:
```suggestion
$message
    ->setSeverity(MessageInterface::SEVERITY_CRITICAL)
    ->setTitle('Unread Critical 1')
    ->save();
```

File: app/code/Magento/AdminNotification/Test/Integration/_files/notifications.php [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+$message->setSeverity(MessageInterface::SEVERITY_MAJOR)
+    ->setTitle('Unread Major 1')
+    ->save();
+
+$message = $om->create(Inbox::class);
+$message->setSeverity(
+    MessageInterface::SEVERITY_CRITICAL
+)->setTitle(
+    'Unread Critical 2'
+)->save();
```
Comment (ihor-sviziev) [inline app/code/Magento/AdminNotification/Test/Integration/_files/notifications.php]:
```suggestion
$message
    ->setSeverity(MessageInterface::SEVERITY_CRITICAL)
    ->setTitle('Unread Critical 2')
    ->save();
```

File: app/code/Magento/AdminNotification/Test/Integration/_files/notifications.php [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+$message = $om->create(Inbox::class);
+$message->setSeverity(
+    MessageInterface::SEVERITY_CRITICAL
+)->setTitle(
+    'Unread Critical 3'
+)->save();
```
Comment (ihor-sviziev) [inline app/code/Magento/AdminNotification/Test/Integration/_files/notifications.php]:
```suggestion
$message
    ->setSeverity(MessageInterface::SEVERITY_CRITICAL)
    ->setTitle('Unread Critical 3')
    ->save();
```

File: app/code/Magento/AdminNotification/Test/Integration/_files/notifications.php [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+$message = $om->create(Inbox::class);
+$message->setSeverity(
+    MessageInterface::SEVERITY_CRITICAL
+)->setTitle(
+    'Read Critical 1'
+)->setIsRead(
+    1
+)->save();
```
Comment (ihor-sviziev) [inline app/code/Magento/AdminNotification/Test/Integration/_files/notifications.php]:
```suggestion
$message
    ->setSeverity(MessageInterface::SEVERITY_CRITICAL)
    ->setTitle('Read Critical 1')
    ->setIsRead(1)
    ->save();
```

File: app/code/Magento/AdminNotification/Test/Integration/_files/notifications.php [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    ->save();
+
+$message = $om->create(Inbox::class);
+$message->setSeverity(
+    MessageInterface::SEVERITY_CRITICAL
+)->setTitle(
+    'Removed Critical 1'
+)->setIsRemove(
+    1
+)->save();
```
Comment (ihor-sviziev) [inline app/code/Magento/AdminNotification/Test/Integration/_files/notifications.php]:
```suggestion
$message
    ->setSeverity(MessageInterface::SEVERITY_CRITICAL)
    ->setTitle('Removed Critical 1')
    ->setIsRemove(1)
    ->save();
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/AdminNotification/Test/Integration/Controller/Adminhtml/Notification/MassMarkAsReadTest.php [hunk 11]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+declare(strict_types=1);
+
+/***
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
```
Comment (ihor-sviziev) [inline app/code/Magento/AdminNotification/Test/Integration/Controller/Adminhtml/Notification/MassMarkAsReadTest.php]:
```suggestion
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */
```

File: app/code/Magento/AdminNotification/Test/Integration/Controller/Adminhtml/Notification/RemoveTest.php [hunk 12]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+declare(strict_types=1);
+
+/***
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
```
Comment (ihor-sviziev) [inline app/code/Magento/AdminNotification/Test/Integration/Controller/Adminhtml/Notification/RemoveTest.php]:
```suggestion
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */
```

File: app/code/Magento/AdminNotification/Test/Integration/Controller/Adminhtml/Notification/MarkAsReadTest.php [hunk 13]
```diff
@@ -1,21 +1,26 @@
 <?php
+
+declare(strict_types=1);
+
 /***
  * Copyright © Magento, Inc. All rights reserved.
  * See COPYING.txt for license details.
  */
```
Comment (ihor-sviziev) [inline app/code/Magento/AdminNotification/Test/Integration/Controller/Adminhtml/Notification/MarkAsReadTest.php]:
```suggestion
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @bartoszkubicki,
Looks good, but seems like some files in dev/tests/integration were not deleted after merging with 2.4-develop branch, and static tests complains about these files. You can find them here:
https://github.com/magento/magento2/tree/352b9efe5bfaafa444c9f15f0ce13b02df1547b2/dev/tests/integration/testsuite/Magento/AdminNotification

Could you remove them?

**Comment by ihor-sviziev** [comment]
@bartoszkubicki  Also would be good to squash all your changes into single commit and force push, if you don't mind.

**Comment by bartoszkubicki** [comment]
@ihor-sviziev ok, tests removed. Should I squash all commits including merge commits?

**Comment by ihor-sviziev** [comment]
@bartoszkubicki yeah would be nice 👍

**Comment by ihor-sviziev** [comment]
@magento run Static Tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
@bartoszkubicki,
Now changes looks really good.
Could you squash all your changes into single commit and I'll approve it?

**Comment by bartoszkubicki** [comment]
@ihor-sviziev I was trying to do it, but unfortunately you have merged 2.4-develop to my branch, and now I am not able to rebase it (i have hundreds of commits nad conflicts). I was trying to work around it, by applying diffs, but with no success. If there is any easy way to do it, please tell me. If not the easiest way to achieve the same effects is to create separate branch and commit there changes without this mess.

**Comment by ihor-sviziev** [comment]
@bartoszkubicki in this case I think you could do [force push](https://evilmartians.com/chronicles/git-push---force-and-how-to-deal-with-it) OR just create new PR from your new branch and close this one.

**Pull request closed without merging**